### PR TITLE
Provide CCW class support via the same pattern as our RCW support

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This PR is an alternative to https://github.com/AaronRobinsonMSFT/ComObjectRedux/pull/6.

This makes the support pluggable in the same manner as our interface support through a new method on the strategy for discovering interface details. It also generates the code in a manner that is better for trimming (the linker and ILC won't remove branches like `if (obj.GetType() == typeof(X))` if `X` is never instantiated as far as I know).

This design ends up using some additional reflection as CCW construction will require one more custom attribute lookup with this model than the alternative, but it also ensures that we don't need to generate checks for any interfaces the user implements that we know are not exposed to COM.

### CODE REVIEW

File: Com.cs [hunk 1]
```diff
@@ -120,6 +151,13 @@ public interface IIUnknownInterfaceDetailsStrategy
     /// <param name="type">RuntimeTypeHandle instance</param>
     /// <returns>Details if type is known.</returns>
     IIUnknownDerivedDetails? GetIUnknownDerivedDetails(RuntimeTypeHandle type);
+
+    /// <summary>
+    /// Given a <see cref="RuntimeTypeHandle"/> get the details about the type that are exposed to COM.
+    /// </summary>
+    /// <param name="type">RuntimeTypeHandle instance</param>
+    /// <returns>Details if type is known.</returns>
+    IComExposedDetails? GetComExposedTypeDetails(RuntimeTypeHandle type);
```
Comment (AaronRobinsonMSFT) [inline Com.cs:160]: We likely need to update the name of this interface.

**Comment by AaronRobinsonMSFT** [comment]
I prefer this model to that of #6. The simplification of `ComputeVtables` is what won me over. I'm not convinced on naming of anything but the mechanics are better. My finding about the `file` scope is rather annoying though. Heavily mangled names makes the entire diagnostic space very cumbersome. We might want to loop with our VS people about this - debugger/profiler.

**Review by jtschuster — APPROVED** [review]
I also prefer this one to #6. It seems better separated with each generated COM class getting its own file, attributes are applied the same way as GeneratedComInterface, and looks like it's more trimmable.

**Comment by jkoritzinsky** [comment]
I’ll open an API proposal for this design instead of #6 and get us on the schedule.

File: Generated/A.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        get
+        {
+            if (_entries == null)
+            {
+                var entries = (ComWrappers.ComInterfaceEntry*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(A), sizeof(ComWrappers.ComInterfaceEntry) * 1);
+                var details = StrategyBasedComWrappers.DefaultIUnknownInterfaceDetailsStrategy.GetIUnknownDerivedDetails(typeof(IComInterface1).TypeHandle)!;
+                entries[0].IID = details.Iid;
+                entries[0].Vtable = (nint)details.ManagedVirtualMethodTable;
+                _entries = entries;
```
Comment (GrabYourPitchforks) [inline Generated/A.cs:21]: Writes need to be volatile since this is a fully unmanaged type and the GC's card table isn't involved.

**Pull request merged**